---
title: "BSEA control vs IBD in cMD for aerophilicity attributes"
output:
  html_document:
    toc: true
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  message = FALSE,
  warning = FALSE
)
```

```{r packages, message=FALSE}
library(bugphyzzAnalyses)
library(bugphyzz)
library(curatedMetagenomicData)
library(dplyr)
library(EnrichmentBrowser)
library(DT)
library(purrr)
library(tibble)
```

# Expected result

+ Enrichment of anaerobic bacteria in healthy samples (gut).
+ Enrichment of facultative anaerobic bacteria in IBD samples (gut).

# Data

Use the HMP_2019_bidmdb study in cMD:

```{r import tse, message=FALSE, warning=FALSE}
dataset_name <- "HallAB_2017.relative_abundance"
tse <- curatedMetagenomicData(
    pattern = dataset_name, 
    dryrun = FALSE, rownames = 'NCBI',
    counts = TRUE
)[[1]]
colData(tse)$GROUP <- ifelse(colData(tse)$study_condition == 'control', 0, 1)
## Filter taxa with at least 1 count as abundance in at least 20% of samples
min_n_samples <- round(ncol(tse) * 0.2)
tse_subset <- tse[rowSums(assay(tse) >= 1) >= min_n_samples,]
tse_subset
```
# bugphyzz data

```{r bp signatures species, message=FALSE}
bp <- importBugphyzz()
aer <- bp$aerophilicity
sp_sigs <- makeSignatures(dat = aer, tax_id_type = 'NCBI_ID', tax_level = 'species') |> 
    discard(is.null)
```

# Differential abundance

DA:

```{r}
assay(tse_subset) <- assay(tse_subset) + 1
edger <- deAna(
    expr = tse_subset, de.method = 'edgeR', padj.method = 'BH',
    filter.by.expr = FALSE 
)
dim(edger)
```

# Enrichment (GSEA) - bugphyzz:

```{r, message=FALSE, warning=FALSE}
edger <- limmaVoom(edger)
gsea <- sbea(
    method = 'gsea', se = edger, gs = sp_sigs, alpha = 0.1, perm = 1000,
    padj.method = 'BH'
)
gsea_res <- as.data.frame(gsea$res.tbl)

caption1 <- c(
  'Table 1. GSEA results comparing the gut microbiome of subjects with
  a normal gut micriobiome (healthy controls) and subjects with IBD.
  ES = Enrichment score, NES = Normalized Enrichment Score. Positive and
  negative scores (ES and NES) indicate enrichment in the IBD and
  healthy control samples, respectively. Bug sets marked with an asterik (*) 
  indicate significant enrichment results (adjusted p-value < 0.1). 
  Benjamini & Hochberg (BH) was used for P value adjustment.'
)

datatable(
  data = gsea_res,
  filter = 'top',
  rownames = FALSE,
  extensions = 'Buttons',
  options = list(
    dom = 'Bft',
    buttons = list('copy', 'print'),
    iDisplayLength = 10,
    keys = TRUE,
    autoWidth = TRUE
  ),
  caption = htmltools::tags$caption(
    style = 'caption-side: top; text-align: left;',
    caption1
  )
)
```

# Import  biosis data

```{r}
biosis <- importBiosis()
row_data <- as.data.frame(rowData(tse_subset))
row_data <- rownames_to_column(row_data, var = 'rowname')
# row_data <- modify_at(row_data, 'genus', ~ as.character(.x))
row_data <- left_join(row_data, biosis, by = c('genus' = 'GenusID'))
biosis_sigs <- split(row_data, factor(row_data$Attribute))
biosis_sigs <- map(biosis_sigs, ~ unique(.x$rowname))
biosis_sigs
```
> No aerobic taxa, which I think should be expected?

# Enrichemnt (GSEA) - biosis

```{r}
gsea2 <- sbea(
    method = 'gsea', se = edger, gs = biosis_sigs, alpha = 0.1, perm = 1000,
    padj.method = 'BH'
)
gsea2_res <- as.data.frame(gsea2$res.tbl)

caption2 <- c(
  'Table 2. GSEA results comparing the gut microbiome of subjects with
  a normal gut micriobiome (healthy controls) and subjects with IBD using
  biosis (nychanes) signatures.
  ES = Enrichment score, NES = Normalized Enrichment Score. Positive and
  negative scores (ES and NES) indicate enrichment in the IBD and
  healthy control samples, respectively. Bug sets marked with an asterik (*) 
  indicate significant enrichment results (adjusted p-value < 0.1). 
  Benjamini & Hochberg (BH) was used for P value adjustment.'
)

datatable(
  data = gsea2_res,
  filter = 'top',
  rownames = FALSE,
  extensions = 'Buttons',
  options = list(
    dom = 'Bft',
    buttons = list('copy', 'print'),
    iDisplayLength = 10,
    keys = TRUE,
    autoWidth = TRUE
  ),
  caption = htmltools::tags$caption(
    style = 'caption-side: top; text-align: left;',
    caption2
  )
)

```

# Session Information

```{r}
sessionInfo()
```
