---
title: "BSEA gingival plaque - MicrobiomeBenchmarkData - 35S"
output:
  html_document:
    toc: true
---

```{r include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  warning = FALSE,
  message = FALSE
)
```

```{r packages, message=FALSE, warning=FALSE}
library(bugphyzzAnalyses)
library(MicrobiomeBenchmarkData)
library(bugphyzz)
library(EnrichmentBrowser)
library(mia)
library(dplyr)
library(DT)
library(purrr)
library(tibble)
library(tidyr)
library(ggplot2)
```

## Data

```{r import tse data, message=FALSE}
dat_name <- 'HMP_2012_16S_gingival_V35'
tse <- getBenchmarkData(dat_name, dryrun = FALSE)[[1]]
tse_genus <- splitByRanks(tse)$genus
## Filter taxa with at least 1 count as abundance in at least 20% of samples
min_n_samples <- round(ncol(tse_genus) * 0.2)
tse_subset <- tse_genus[rowSums(assay(tse_genus) >= 1) >= min_n_samples,]
tse_subset
```

Number of samples per condition

```{r}
table(tse_subset$body_subsite)
```

## bugphyzz signatures

Signatures at the genus level:

```{r bp signatures genus, message = FALSE}
discrete_types <- c(
    "binary", "multistate-intersection", "multistate-union"
)
bp <- importBugphyzz(v = 0.5) |> 
    map(~ mutate(.x, NCBI_ID = as.character(NCBI_ID))) |> 
    map(~ {
        attr_type <- unique(.x$Attribute_type)
        if (attr_type %in% discrete_types) {
            df <- .x |> 
                filter(
                    !(Validation < 0.7 & Evidence == "asr")
                )
        } else if (attr_type == "numeric") {
            df <- .x |> 
                filter(
                    !(Validation < 0.5 & Evidence == "asr")
                )
        }
        df
    })
aer <- bp$aerophilicity
gn_sigs <- makeSignatures(
    dat = aer, tax_id_type = 'Taxon_name', tax_level = 'genus'
)
## Only one genus is missing  (TG5), which is ambiguous in the ncbi taxonomy
mean(as.integer(
    rownames(tse_subset) %in% unique(unlist(gn_sigs, use.names = FALSE ))
))
```

Without ASR some annotations are missing:

```{r}
gn_sigs_noasr <- makeSignatures(
    dat = aer, tax_id_type = 'Taxon_name', tax_level = 'genus',
    evidence = c("exp", "igc", "tas", "nas", "tax")
)
mean(as.integer(
    rownames(tse_subset) %in% unique(unlist(gn_sigs_noasr, use.names = FALSE))
))
```

## Differential abundance

```{r}
tse_subset$GROUP <- ifelse(
  tse_subset$body_subsite == 'subgingival_plaque', 0, 1
)
edger <- deAna(
    expr = tse_subset, de.method = 'edgeR', padj.method = 'fdr', 
    filter.by.expr = FALSE, 
)
edger <- limmaVoom(edger)
```

## Enrichment (GSEA)

-   `*` significant (Pvalue < 0.05)
-   `+` enriched in supragingival plaque
-   `-` enriched in subgingival plaque

```{r, message=FALSE}
gsea <- sbea(
  method = 'gsea', se = edger, gs = gn_sigs, perm = 1000,
  alpha = 0.1 # This is for DA taxa
)
gsea_tbl <- as.data.frame(gsea$res.tbl) |> 
  mutate(
    GENE.SET = ifelse(PVAL < 0.05, paste0(GENE.SET, ' *'), GENE.SET),
    PVAL = round(PVAL, 3),
  ) |> 
  rename(BUG.SET = GENE.SET)
myDataTable(gsea_tbl)
```

## Enrichment (GSEA) no asr

Even without ASR annotations, the  enrichment results are still the same.

```{r}
gsea <- sbea(
  method = 'gsea', se = edger, gs = gn_sigs_noasr, perm = 1000,
  alpha = 0.1 # This is for DA taxa
)
gsea_tbl <- as.data.frame(gsea$res.tbl) |> 
  mutate(
    GENE.SET = ifelse(PVAL < 0.05, paste0(GENE.SET, ' *'), GENE.SET),
    PVAL = round(PVAL, 3),
  ) |> 
  rename(BUG.SET = GENE.SET)
myDataTable(gsea_tbl)
```

## Enrichment (ORA)

Combine into a single table:

```{r}
edger_up <- edger
gn_sigs_bk <- c(gn_sigs, list(background = rownames(edger)))
## Ora is performed by comparing the significant group (p.adjust < 0.1)
## against the background. The background should include all of the
## taxa in the experiment. That's why I assigned an adjusted p-value of 
## 1 to all taxa with a negative FC in the code below.
rowData(edger_up)$ADJ.PVAL <- ifelse(
  rowData(edger_up)$FC < 0, 1, rowData(edger_up)$ADJ.PVAL
)
ora_up <- sbea(
  method = 'ora', se = edger_up, gs = gn_sigs_bk, perm = 0,
  alpha = 0.1 # This is for DA taxa
)
ora_up_df <- as.data.frame(ora_up$res.tbl)
ora_up_df$BODY.SITE <- 'Supragingival plaque'

edger_down <- edger
rowData(edger_down)$ADJ.PVAL <- ifelse(
  rowData(edger_down)$FC > 0, 1, rowData(edger_down)$ADJ.PVAL
)
ora_down <- sbea(
  method = 'ora', se = edger_down, gs = gn_sigs_bk, perm = 0, 
  alpha = 0.1,
)
ora_down_df <- data.frame(ora_down$res.tbl)
ora_down_df$BODY.SITE <- 'Subgingival plaque'

ora <- bind_rows(ora_up_df, ora_down_df) |> 
  arrange(BODY.SITE, PVAL) |> 
  relocate(BODY.SITE) |> 
  mutate(
    GENE.SET = ifelse(PVAL < 0.05, paste0(GENE.SET, ' *'), GENE.SET),
    PVAL = round(PVAL, 3),
  ) |> 
  dplyr::rename(
    BUG.SET = GENE.SET, NR.BUGS = NR.GENES, NR.SIG.BUGS = NR.SIG.GENES
  ) |> 
    filter(BUG.SET != "background")
myDataTable(ora)
```

## dbBact approach

```{r, warning=FALSE}
bpSigs_g <- map(
    bp, ~ {
        makeSignatures(.x, tax_id_type = "NCBI_ID", tax_level = "genus")
}) |> 
    list_flatten(name_spec = "{inner}") |> 
    discard(is.null)

# bpSigs_g <- bpSigs_g[grep("aerophilicity", names(bpSigs_g))]
```


```{r}
con <- rownames(edger)[which(rowData(edger)$FC > 0 & rowData(edger)$ADJ.PVAL < 0.1)]
con <- taxizedb::name2taxid(con, db = "ncbi", out_type = "summary")
con <- con$id[con$id != "508215"]
attr(con, "nexp") <- 1
```


```{r}
cas <- rownames(edger)[which(rowData(edger)$FC < 0 & rowData(edger)$ADJ.PVAL < 0.1)]
cas <- cas[which(cas != "TG5")]
cas <- taxizedb::name2taxid(cas, db = "ncbi")
attr(cas, "nexp") <- 1

```


```{r}
se <- dbEn2(con, cas, bpSigs_g)
```

```{r, fig.width = 12, fig.height=12}
dbHt(se)
```
# ORA - NO ASR

```{r}
gn_sigs_noasr_bk <- c(gn_sigs_noasr, list(background = rownames(edger)))
ora_up_noasr <- sbea(
  method = 'ora', se = edger_up, gs = gn_sigs_noasr_bk, perm = 0, 
  alpha = 0.1 # This is for DA taxa
)
ora_up_df_noasr <- as.data.frame(ora_up_noasr$res.tbl)
ora_up_df_noasr$BODY.SITE <- 'Supragingival plaque'

ora_down_noasr <- sbea(
  method = 'ora', se = edger_down, gs = gn_sigs_noasr_bk, perm = 0, 
  alpha = 0.1 # This is for DA taxa
)
ora_down_df_noasr <- data.frame(ora_down_noasr$res.tbl)
ora_down_df_noasr$BODY.SITE <- 'Subgingival plaque'

ora_noasr <- bind_rows(ora_up_df_noasr, ora_down_df_noasr) |> 
  arrange(BODY.SITE, PVAL) |> 
  relocate(BODY.SITE) |> 
  mutate(
    GENE.SET = ifelse(PVAL < 0.05, paste0(GENE.SET, ' *'), GENE.SET),
    PVAL = round(PVAL, 3),
  ) |> 
  dplyr::rename(
    BUG.SET = GENE.SET, NR.BUGS = NR.GENES, NR.SIG.BUGS = NR.SIG.GENES
  ) |> 
    filter(BUG.SET != "background")

myDataTable(ora_noasr)
```

# Quick checks

Abundance of taxa in the 'down-regulated' set:

```{r}
## Couldn't make tidySummarizedExperiment work.
## Update code later.
counts1 <- assay(edger_down, 'counts')
counts1 <- log2(counts1 + 1)
counts1 <- counts1 |> 
  as.data.frame() |> 
  rownames_to_column(var = 'taxname') |> 
  pivot_longer(
    cols = 2:last_col(), names_to = 'sample', values_to = 'abundance'
  )
row_data1 <- edger_down |> 
  rowData() |> 
  as.data.frame() |> 
  rownames_to_column(var = 'taxname') |> 
  as_tibble()
col_data1 <- edger_down |> 
  colData() |> 
  as.data.frame() |> 
  rownames_to_column(var = 'sample') |> 
  as_tibble()
data1 <- left_join(counts1, row_data1, by = 'taxname') |> 
  left_join(col_data1, by = 'sample')

p1 <- data1 |> 
  filter(ADJ.PVAL < 0.1) |> 
  ggplot(aes(body_subsite, abundance)) +
  labs(
    title = 'Down-regulated supragingival = Up-regulated in subgingival'
  ) +
  geom_boxplot() +
  geom_point()
p1
```

Abundance of taxa in the 'up-regulated' set:

```{r}
## Couldn't make tidySummarizedExperiment work.
## Update code later.
counts2 <- assay(edger_up, 'counts')
counts2 <- log2(counts2 + 1)
counts2 <- counts2 |> 
  as.data.frame() |> 
  rownames_to_column(var = 'taxname') |> 
  pivot_longer(
    cols = 2:last_col(), names_to = 'sample', values_to = 'abundance'
  )
row_data2 <- edger_up |> 
  rowData() |> 
  as.data.frame() |> 
  rownames_to_column(var = 'taxname') |> 
  as_tibble()
col_data2 <- edger_up |> 
  colData() |> 
  as.data.frame() |> 
  rownames_to_column(var = 'sample') |> 
  as_tibble()
data2 <- left_join(counts2, row_data2, by = 'taxname') |> 
  left_join(col_data2, by = 'sample')

p2 <- data2 |> 
  filter(ADJ.PVAL < 0.1) |> 
  ggplot(aes(body_subsite, abundance)) +
  labs(
    title = 'Up-regulated supragingival = Down-regulated in subgingival'
  ) +
  geom_boxplot() +
  geom_point()
p2
```

# Session information

```{r session info}
sessioninfo::session_info()
```
